rust/arrow/src/compute/cast_kernels.rs

```diff
@@ -0,0 +1,543 @@
+//!
+//! * Boolean to Utf8: `true` => '1', `false` => `0`
+//! * Utf8 to numeric: strings that can't be parsed to numbers return null, float strings
+//!   in integer casts return null
+//! * Numeric to boolean: 0 returns `false`, any other value returns `true`
+//!
+//! ## Unsupported Casts
+//!
+//! * To or from `StructArray`
+//! * To or from `ListArray`
+//! * Boolean to float
+//! * Utf8 to boolean
+//!
+//! Example:
+//!
+//! ```
+//! use arrow::array::*;
+//! use arrow::compute::cast;
+//! use arrow::datatypes::DataType;
+//! use std::sync::Arc;
+//!
+//! let a = Int32Array::from(vec![5, 6, 7]);
+//! let array = Arc::new(a) as ArrayRef;
+//! let b = cast(&array, &DataType::Float64).unwrap();
+//! let c = b.as_any().downcast_ref::<Float64Array>().unwrap();
+//! assert_eq!(5.0, c.value(0));
+//! assert_eq!(6.0, c.value(1));
+//! assert_eq!(7.0, c.value(2));
+//! ```
+
+use std::sync::Arc;
+
+use crate::array::*;
+use crate::builder::*;
+use crate::datatypes::*;
+use crate::error::{ArrowError, Result};
+
+/// Macro rule to cast between numeric types
+macro_rules! cast_numeric_arrays {
+    ($array:expr, $from_ty:ident, $to_ty:ident) => {{
+        match numeric_cast::<$from_ty, $to_ty>(
+            $array
+                .as_any()
+                .downcast_ref::<PrimitiveArray<$from_ty>>()
+                .unwrap(),
+        ) {
+            Ok(to) => Ok(Arc::new(to) as ArrayRef),
+            Err(e) => Err(e),
+        }
+    }};
+}
+
+macro_rules! cast_numeric_to_string {
+    ($array:expr, $from_ty:ident) => {{
+        match cast_numeric_to_string::<$from_ty>(
+            $array
+                .as_any()
+                .downcast_ref::<PrimitiveArray<$from_ty>>()
+                .unwrap(),
+        ) {
+            Ok(to) => Ok(Arc::new(to) as ArrayRef),
+            Err(e) => Err(e),
+        }
+    }};
+}
+
+macro_rules! cast_string_to_numeric {
+    ($array:expr, $to_ty:ident) => {{
+        match cast_string_to_numeric::<$to_ty>(
+            $array.as_any().downcast_ref::<BinaryArray>().unwrap(),
+        ) {
+            Ok(to) => Ok(Arc::new(to) as ArrayRef),
+            Err(e) => Err(e),
+        }
+    }};
+}
+
+macro_rules! cast_numeric_to_bool {
+    ($array:expr, $from_ty:ident) => {{
+        match cast_numeric_to_bool::<$from_ty>(
+            $array
+                .as_any()
+                .downcast_ref::<PrimitiveArray<$from_ty>>()
+                .unwrap(),
+        ) {
+            Ok(to) => Ok(Arc::new(to) as ArrayRef),
+            Err(e) => Err(e),
+        }
+    }};
+}
+
+macro_rules! cast_bool_to_numeric {
+    ($array:expr, $to_ty:ident) => {{
+        match cast_bool_to_numeric::<$to_ty>(
+            $array.as_any().downcast_ref::<BooleanArray>().unwrap(),
+        ) {
+            Ok(to) => Ok(Arc::new(to) as ArrayRef),
+            Err(e) => Err(e),
+        }
+    }};
+}
+
+/// Cast array to provided data type
+pub fn cast(array: &ArrayRef, to_type: &DataType) -> Result<ArrayRef> {
```

Comment:
can we move this to the top and add more comments on its behavior?

Reply:
I can't put it at the top as it depends on the macro definitions. I've expanded the doc comment to include behaviour.

Reply:
Resolved this by removing macros